Mathlib/Topology/Instances/ENNReal.lean

```diff
@@ -1361,6 +1361,32 @@ lemma liminf_const_sub (F : Filter ι) [NeBot F] (f : ι → ℝ≥0∞) {c : 
   (Antitone.map_limsSup_of_continuousAt (F := F.map f) (f := fun (x : ℝ≥0∞) ↦ c - x)
     (fun _ _ h ↦ tsub_le_tsub_left h c) (continuous_sub_left c_ne_top).continuousAt).symm
 
+lemma mul_liminf_le_liminf_mul {α : Type*} {f : Filter α} {u v : α → ℝ≥0∞} :
+    (liminf u f) * (liminf v f) ≤ liminf (u * v) f := by
+  refine mul_le_of_forall_mul_le fun a a_u b b_v ↦ (le_liminf_iff).2 fun c c_ab ↦ ?_
+  filter_upwards [eventually_lt_of_lt_liminf a_u, eventually_lt_of_lt_liminf b_v] with x a_x b_x
+  exact c_ab.trans (mul_lt_mul a_x b_x)
+
+lemma limsup_mul_le_mul_limsup {α : Type*} {f : Filter α} {u v : α → ℝ≥0∞}
+    (h : limsup u f ≠ 0 ∨ limsup v f ≠ ∞) (h' : limsup u f ≠ ∞ ∨ limsup v f ≠ 0) :
+    limsup (u * v) f ≤ (limsup u f) * (limsup v f) := by
+  refine le_mul_of_forall_le_mul h h' fun a a_u b b_v ↦ (limsup_le_iff).2 fun c c_ab ↦ ?_
+  filter_upwards [eventually_lt_of_limsup_lt a_u, eventually_lt_of_limsup_lt b_v] with x a_x b_x
+  exact (mul_lt_mul a_x b_x).trans c_ab
+
+lemma limsup_mul_liminf_le_limsup_mul {α : Type*} {f : Filter α} {u v : α → ℝ≥0∞} :
+    (limsup u f) * (liminf v f) ≤ limsup (u * v) f :=
+  mul_le_of_forall_mul_le fun a a_u b b_v ↦ (le_limsup_iff).2 fun c c_ab ↦
+    Frequently.mono (Frequently.and_eventually ((frequently_lt_of_lt_limsup) a_u)
+    ((eventually_lt_of_lt_liminf) b_v)) fun _ ab_x ↦ c_ab.trans (mul_lt_mul ab_x.1 ab_x.2)
+
+lemma liminf_mul_le_limsup_mul_liminf {α : Type*} {f : Filter α} {u v : α → ℝ≥0∞}
+    (h : limsup u f ≠ 0 ∨ liminf v f ≠ ∞) (h' : limsup u f ≠ ∞ ∨ liminf v f ≠ 0) :
+    liminf (u * v) f ≤ (limsup u f) * (liminf v f) :=
+  le_mul_of_forall_le_mul h h' fun a a_u b b_v ↦ (liminf_le_iff).2 fun c c_ab ↦
+    Frequently.mono (Frequently.and_eventually ((frequently_lt_of_liminf_lt) b_v)
+    ((eventually_lt_of_limsup_lt) a_u)) fun _ ab_x ↦ (mul_lt_mul ab_x.2 ab_x.1).trans c_ab
```

Comment:
This one looks too complicated as well. Can you try starting the proof with 
```suggestion
    liminf (u * v) f ≤ (limsup u f) * (liminf v f) :=
  sSup_le fun i hi ↦ _
```

Reply:
I don't see how to make it work. What do you have in mind after that?

The problem is that I get something like ` ∀ᶠ (a : α) in f, x ≤ u a * v a`, but then, you need to "split" `u a` and `v a` into a limsup and a liminf. Note also that the two hypotheses `h` and `h'` are necessary here.

Reply:
Damn I need to think 😅 

Will sleep over it, and hopefully tomorrow comes with an idea

Reply:
I had no time until now. Will look into it